File: components/permissions/src/main/java/org/cloudfoundry/credhub/requests/PermissionEntry.java
package org.cloudfoundry.credhub.requests;


import java.util.List;

import javax.validation.constraints.NotEmpty;

import org.springframework.validation.annotation.Validated;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.collect.Lists;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.cloudfoundry.credhub.ErrorMessages;
import org.cloudfoundry.credhub.PermissionOperation;

@JsonAutoDetect
@Validated
public class PermissionEntry {
  @NotEmpty(message = ErrorMessages.Permissions.MISSING_ACTOR)
  private String actor;

  @NotEmpty(message = ErrorMessages.Permissions.MISSING_PATH)
  private String path;

  @NotEmpty(message = ErrorMessages.Permissions.MISSING_OPERATIONS)
  @JsonProperty("operations")
  private List<PermissionOperation> allowedOperations;

  public PermissionEntry() {
    super();
  }

  public PermissionEntry(final String actor, final String path, final PermissionOperation... operations) {
    this(actor, path, Lists.newArrayList(operations));
  }

  public PermissionEntry(final String actor, final String path, final List<PermissionOperation> operations) {
    super();
    this.actor = actor;
    this.path = path;
    this.allowedOperations = operations;
  }

  public String getActor() {
    return actor;
  }

  public void setActor(final String actor) {
    this.actor = actor;
  }

  public List<PermissionOperation> getAllowedOperations() {
    return allowedOperations;
  }

  public void setAllowedOperations(final List<PermissionOperation> allowedOperations) {
    this.allowedOperations = allowedOperations;
  }

  public String getPath() {
    return path;
  }

  public void setPath(final String path) {
    this.path = path;
  }

  @Override
  public boolean equals(final Object o) {
    if (this == o) {
      return true;
    }

    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    final PermissionEntry that = (PermissionEntry) o;

    return new EqualsBuilder()
      .append(actor, that.actor)
      .append(path, that.path)
      .append(allowedOperations, that.allowedOperations)
      .isEquals();
  }

  @Override
  public int hashCode() {
    return new HashCodeBuilder(17, 37)
      .append(actor)
      .append(path)
      .append(allowedOperations)
      .toHashCode();
  }
}

File: components/permissions/src/main/kotlin/org/cloudfoundry/credhub/requests/PermissionEntry.kt
package org.cloudfoundry.credhub.requests

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.JsonProperty
import com.google.common.collect.Lists
import jakarta.validation.constraints.NotEmpty
import org.apache.commons.lang3.builder.EqualsBuilder
import org.apache.commons.lang3.builder.HashCodeBuilder
import org.cloudfoundry.credhub.ErrorMessages
import org.cloudfoundry.credhub.PermissionOperation
import org.springframework.validation.annotation.Validated

@JsonAutoDetect
@Validated
class PermissionEntry {
    @NotEmpty(message = ErrorMessages.Permissions.MISSING_ACTOR)
    var actor: String? = null

    @NotEmpty(message = ErrorMessages.Permissions.MISSING_PATH)
    var path: String? = null

    @NotEmpty(message = ErrorMessages.Permissions.MISSING_OPERATIONS)
    @JsonProperty("operations")
    var allowedOperations: List<PermissionOperation>? = null

    constructor() : super() {}

    constructor(actor: String, path: String, vararg operations: PermissionOperation) : this(
        actor,
        path,
        Lists.newArrayList<PermissionOperation>(*operations),
    ) {
    }

    constructor(actor: String, path: String?, operations: List<PermissionOperation>) : super() {
        this.actor = actor
        this.path = path
        this.allowedOperations = operations
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        if (other == null || javaClass != other.javaClass) {
            return false
        }

        val that = other as PermissionEntry?

        return EqualsBuilder()
            .append(actor, that!!.actor)
            .append(path, that.path)
            .append(allowedOperations, that.allowedOperations)
            .isEquals
    }

    override fun hashCode(): Int =
        HashCodeBuilder(17, 37)
            .append(actor)
            .append(path)
            .append(allowedOperations)
            .toHashCode()
}
